File: app/src/main/java/io/github/hrilab/pepper_realtime/tools/entertainment/PlayYouTubeVideoTool.java
package io.github.hrilab.pepper_realtime.tools.entertainment;

import android.util.Log;
import io.github.hrilab.pepper_realtime.YouTubeSearchService;
import io.github.hrilab.pepper_realtime.tools.Tool;
import io.github.hrilab.pepper_realtime.tools.ToolContext;

import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Tool for searching and playing YouTube videos.
 * Uses YouTube Data API to find videos and displays them in the UI.
 */
public class PlayYouTubeVideoTool implements Tool {
    
    private static final String TAG = "PlayYouTubeVideoTool";

    @Override
    public String getName() {
        return "play_youtube_video";
    }

    @Override
    public JSONObject getDefinition() {
        try {
            JSONObject tool = new JSONObject();
            tool.put("type", "function");
            tool.put("name", getName());
            tool.put("description", "Search and play a YouTube video based on user's request. Use this when the user asks to play music, songs, or videos from YouTube.");
            
            JSONObject params = new JSONObject();
            params.put("type", "object");
            
            JSONObject properties = new JSONObject();
            properties.put("query", new JSONObject()
                .put("type", "string")
                .put("description", "Search query for the video (e.g. 'like a virgin madonna', 'funny cat videos')"));
            
            params.put("properties", properties);
            params.put("required", new JSONArray().put("query"));
            tool.put("parameters", params);
            
            return tool;
        } catch (Exception e) {
            throw new RuntimeException("Failed to create tool definition", e);
        }
    }

    @Override
    public String execute(JSONObject args, ToolContext context) throws Exception {
        String query = args.optString("query", "");
        
        // Validate required parameters
        if (query.isEmpty()) {
            return new JSONObject().put("error", "Missing required parameter: query").toString();
        }
        
        // Check if YouTube API is available
        if (!context.getApiKeyManager().isYouTubeAvailable()) {
            String setupMessage = context.getApiKeyManager().getYouTubeSetupMessage();
            return new JSONObject().put("error", "YouTube API key not configured").put("setup", setupMessage).toString();
        }
        
        Log.i(TAG, "Searching YouTube for: " + query);
        
        try {
            // Create YouTube search service
            YouTubeSearchService youtubeService = new YouTubeSearchService(context.getApiKeyManager().getYouTubeApiKey());
            
            // Search for video
            YouTubeSearchService.YouTubeVideo video = youtubeService.searchFirstVideo(query);
            
            if (video == null) {
                return new JSONObject().put("error", "No videos found for query: " + query).toString();
            }
            
            // Show video player via manager on main thread
            if (context.hasUi()) {
                context.getActivity().runOnUiThread(() -> io.github.hrilab.pepper_realtime.managers.YouTubePlayerManager.showYouTubePlayerDialog(
                        context.getActivity(), video,
                        new io.github.hrilab.pepper_realtime.managers.YouTubePlayerManager.YouTubePlayerCallback() {
                            @Override
                            public void onPlayerOpened() {
                                context.sendAsyncUpdate("🎵 YouTube video started - microphone muted", false);
                            }
                            
                            @Override
                            public void onPlayerClosed() {
                                context.sendAsyncUpdate("🎵 YouTube video ended - microphone active again", false);
                            }
                            
                            @Override
                            public boolean isActivityFinishing() {
                                return context.getActivity().isFinishing();
                            }
                        }
                    ));
            }
            
            // Return success response
            JSONObject result = new JSONObject();
            result.put("status", "Video player opened");
            result.put("query", query);
            result.put("video_title", video.getTitle());
            result.put("channel", video.getChannelTitle());
            result.put("video_id", video.getVideoId());
            return result.toString();
            
        } catch (Exception e) {
            Log.e(TAG, "Error playing YouTube video", e);
            return new JSONObject().put("error", "Failed to search or play video: " + e.getMessage()).toString();
        }
    }

    @Override
    public boolean requiresApiKey() {
        return true;
    }

    @Override
    public String getApiKeyType() {
        return "YouTube";
    }
}

File: app/src/main/java/ch/fhnw/pepper_realtime/tools/entertainment/PlayYouTubeVideoTool.kt
package ch.fhnw.pepper_realtime.tools.entertainment

import android.util.Log
import ch.fhnw.pepper_realtime.manager.YouTubePlayerManager
import ch.fhnw.pepper_realtime.service.YouTubeSearchService
import ch.fhnw.pepper_realtime.tools.ApiKeyRequirement
import ch.fhnw.pepper_realtime.tools.ApiKeyType
import ch.fhnw.pepper_realtime.tools.Tool
import ch.fhnw.pepper_realtime.tools.ToolContext
import org.json.JSONArray
import org.json.JSONObject

import kotlinx.coroutines.runBlocking

/**
 * Tool for searching and playing YouTube videos.
 * Uses YouTube Data API to find videos and displays them in the UI.
 */
class PlayYouTubeVideoTool : Tool {

    override fun getName(): String = "play_youtube_video"

    override fun getDefinition(): JSONObject {
        return JSONObject().apply {
            put("type", "function")
            put("name", getName())
            put("description", "Search and play a YouTube video based on user's request. Use this when the user asks to play music, songs, or videos from YouTube.")

            put("parameters", JSONObject().apply {
                put("type", "object")
                put("properties", JSONObject().apply {
                    put("query", JSONObject()
                        .put("type", "string")
                        .put("description", "Search query for the video (e.g. 'like a virgin madonna', 'funny cat videos')"))
                })
                put("required", JSONArray().put("query"))
            })
        }
    }

    override fun execute(args: JSONObject, context: ToolContext): String {
        val query = args.optString("query", "")

        // Validate required parameters
        if (query.isEmpty()) {
            return JSONObject().put("error", "Missing required parameter: query").toString()
        }

        // Check if YouTube API is available
        if (!context.apiKeyManager.isYouTubeAvailable()) {
            val setupMessage = context.apiKeyManager.youTubeSetupMessage
            val result = JSONObject()
            result.put("error", "YouTube API key not configured")
            result.put("setup", setupMessage as String?)
            return result.toString()
        }

        Log.i("PlayYouTubeVideoTool", "Searching YouTube for: $query")

        return try {
            // Create YouTube search service
            val youtubeService = YouTubeSearchService(context.apiKeyManager.youTubeApiKey)

            // Search for video using runBlocking since execute is synchronous
            val video = runBlocking {
                youtubeService.searchFirstVideo(query)
            } ?: return JSONObject().put("error", "No videos found for query: $query").toString()

            // Show video player via manager on main thread
            if (context.hasUi()) {
                val activity = context.activity
                activity?.runOnUiThread {
                    YouTubePlayerManager.showYouTubePlayerDialog(
                        activity,
                        video,
                        object : YouTubePlayerManager.YouTubePlayerCallback {
                            override fun onPlayerOpened() {
                                // Mute microphone while YouTube video is playing
                                context.toolHost.muteMicrophone()
                                context.sendAsyncUpdate("🎵 YouTube video started - microphone muted", false)
                            }

                            override fun onPlayerClosed() {
                                // Unmute microphone when video ends
                                context.toolHost.unmuteMicrophone()
                                context.sendAsyncUpdate("🎵 YouTube video ended - microphone active again", false)
                                // Refresh chat UI to show any updates that were missed while overlay was active
                                context.toolHost.refreshChatMessages()
                            }

                            override fun isActivityFinishing(): Boolean {
                                return activity.isFinishing
                            }
                        }
                    )
                }
            }

            // Return success response
            JSONObject().apply {
                put("status", "Video player opened")
                put("query", query)
                put("video_title", video.title)
                put("channel", video.channelTitle)
                put("video_id", video.videoId)
            }.toString()

        } catch (e: Exception) {
            Log.e("PlayYouTubeVideoTool", "Error playing YouTube video", e)
            JSONObject().put("error", "Failed to search or play video: ${e.message}").toString()
        }
    }

    override val apiKeyRequirement = ApiKeyRequirement.Required(ApiKeyType.YOUTUBE)


}


